src/Thread.tsx

```diff
@@ -0,0 +1,51 @@
+import {defined} from './Base';
+import {fetchMessageIdsAndLabels, fetchMessagesByIds} from './Gapi';
+import {Message} from './Message';
+
+export class Thread {
+  private _rawThread: gapi.client.gmail.Thread;
+  private _messages: Message[] | undefined;
+  private _firstAndLastMessages: Message[] | undefined;
+
+  constructor(thread: gapi.client.gmail.Thread) {
+    this._rawThread = thread;
+    this._messages = undefined;
+    this._firstAndLastMessages = undefined;
+  }
+
+  id(): string {
+    return defined(this._rawThread.id);
+  }
+
+  messages(): Message[] {
+    return defined(this._messages);
+  }
+
+  firstAndLastMessages(): Message[] | undefined {
+    return this._firstAndLastMessages;
+  }
+
+  hasMessagesInInbox(): boolean {
+    return this.messages()?.some((x) => x.labelIds().includes('INBOX'));
+  }
+
+  async fetchMessages(): Promise<void> {
+    this._rawThread = await fetchMessageIdsAndLabels(this.id());
+    const messages = this._rawThread.messages?.map((x) => new Message(x));
+    this._messages = messages;
+
+    const messageIdsToFetch: string[] = [];
+    if (messages?.length) {
+      messageIdsToFetch.push(messages?.[0]?.id());
```

Comment:
messages here doesn't need the optional chaining, right?

Reply:
Does the [0]? need it? Not sure how smart typescript error checking is here, but it's guaranteed to be there given the if statement above I think.